crates/ruff/src/checkers/ast/analyze/deferred_scopes.rs

```diff
@@ -169,15 +169,18 @@ pub(crate) fn deferred_scopes(checker: &mut Checker) {
                         }
 
                         // If this is an overloaded function, abort.
-                        if shadowed.kind.is_function_definition()
-                            && visibility::is_overload(
-                                cast::decorator_list(
-                                    checker.semantic.stmts[shadowed.source.unwrap()],
-                                ),
-                                &checker.semantic,
-                            )
-                        {
-                            continue;
+                        if shadowed.kind.is_function_definition() {
+                            if let Some(source) = shadowed.source {
+                                if let Stmt::FunctionDef(function) = &checker.semantic.stmts[source]
+                                {
+                                    if visibility::is_overload(
+                                        &function.decorator_list,
+                                        &checker.semantic,
+                                    ) {
+                                        continue;
+                                    }
+                                }
+                            }
```

Comment:
Nit: this feels more difficult to understand than what we used to have. Furthermore, the check for `shadowed.kind.is_function_definition` and the `if let Stmt::FunctionDef(...) = ..` seem redundant. Do you see a way to remove the redundancy / simplify the API?

Reply:
They're actually note quite redundant, because you can have bindings whose source statement is a function definition, but that aren't _themselves_ function definitions (for example, function parameters). Trying to clean up a bit.